core/src/main/scala/filodb.core/memstore/ratelimit/CardinalityTracker.scala

```diff
@@ -102,14 +114,106 @@ class CardinalityTracker(ref: DatasetRef,
     }
   }
 
+  /**
+   * Updates the DOWNSAMPLE CLUSTER's cardinality count in the cardinalityCountMap. Flushes the data to RocksDB
+   * when `dsCardinalityMapFlushCount` threshold reached
+   *
+   *  NOTE: The cardinality count increment happens by `1`.
+   *
+   * Cardinality count at each level of shardKey needs to be updated
+   * For example: if shardKey = (my_ws, my_ns, my_metric), then we have to update
+   * the cardinality count of 4 prefixes. They are -
+   * 1. (total across all ws)
+   * 2. (my_ws)
+   * 3. (my_ws, my_ns)
+   * 4. (my_ws, my_ns, my_metric)
+   *
+   * @param shardKey
+   */
+  def updateCardinalityCountsDS(shardKey: Seq[String]): Unit = synchronized {
+    (0 to shardKey.length).foreach { i =>
+
+      // update current prefix's cardinality count
+      val prefix = shardKey.take(i)
+      val cardCountRecord = cardinalityCountMapDS.get(prefix)
+        .map(x => (x._1 + 1, x._2))
+        .getOrElse((1, 0)) // child prefix update parent's childrenCount
+      cardinalityCountMapDS.put(prefix, cardCountRecord)
+
+      // update children count of parent and throw exception if quota reached
+      if (i > 0) {
+        val parentPrefix = shardKey.take(i - 1)
+
+        // we always add parent before the child, hence it is okay to get the parent prefix's record directly
+        // without the None check
+        val updatedCountRecord = cardinalityCountMapDS.get(parentPrefix)
+          .map(x => (x._1, x._2 + 1)).get
+
+        // check if number of children is higher than the given quota. This allows us guard our physical resources
+        // and avoid failures because of runaway cardinality
+        val childrenQuota = defaultChildrenQuota(parentPrefix.length)
+        if (updatedCountRecord._2 > childrenQuota) {
+          quotaExceededProtocol.quotaExceeded(ref, shard, prefix, childrenQuota)
+          throw QuotaReachedException(prefix, prefix, childrenQuota)
+        }
+
+        // store the updated parent's childrenCount
+        cardinalityCountMapDS.put(parentPrefix, updatedCountRecord)
+      }
+    }
+
+    if (cardinalityCountMapDS.size > dsCardinalityMapFlushCount) {
```

Comment:
1. Yeah, the implementation doesn't make any assumption of ordering guarantees and same/keys prefix will most likely happen again. Chose this implementation to avoid any complexity and keep implementation plain and simple ( no sort or other transformation. The reason for threshold and guards here is to avoid runaway use of memory )

2. Yep, I was thinking of calling the flush from the invoker/caller which starts the cardinality count in DS